File: coding_py/boj/bj1867.py
import sys
input = sys.stdin.readline


def dfs(i):
    for want in stone[i]:
        if check[want]:
            continue
        check[want] = True
        if matching[want] == 0 or dfs(matching[want]):
            matching[want] = i
            return True 
    return False 


n, k = map(int, input().split())
stone = [[] for _ in range(n+1)]
matching = [0 for _ in range(n+1)]
res = 0
for _ in range(k):
    i,j = map(int,input().split())
    stone[i].append(j)

for i in range(1, n+1):
    check = [False for _ in range(n+1)] #방문 여부 리스트
    if dfs(i):
        res += 1

print(res)

File: algo_cpp/bj1867.cpp
#include<cstdio>
#include<vector>
#define MAX 501
using namespace std;

int stone[MAX];
bool check[MAX];
vector<int> person[MAX];

bool dfs(int i) {
	for (auto w : person[i]) {
		if (check[w])
			continue;
		check[w] = 1;
		if (stone[w] == 0 || dfs(stone[w])) {
			stone[w] = i;
			return true;
		}
	}
	return false;
}

int main(void)
{
	int n, m, res = 0;
	scanf("%d%d", &n, &m);
	for (int i = 1; i <= m; i++) {
		int a, b;
		scanf("%d%d", &a, &b);
		person[a].push_back(b);
	}
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++)
			check[j] = 0;
		if (dfs(i))
			res++;
	}
	printf("%d", res);
	return 0;
}